src/testutil/mod.rs

```diff
@@ -0,0 +1,276 @@
+//! # Test Utilities
+//!
+//! For now, mostly just helpers for running "ui tests", or executing forth code at
+//! test time.
+//!
+//! ## UI Tests
+//!
+//! Generally, forth code provided as a str will have one of the following things
+//! for each line:
+//!
+//! * Configuration values for the VM, specified as "frontmatter comments".
+//!   These must appear before any other non-comment lines. Currently accepted:
+//!     * `( data_stack_elems USIZE )`
+//!     * `( return_stack_elems USIZE )`
+//!     * `( control_stack_elems USIZE )`
+//!     * `( input_buf_elems USIZE )`
+//!     * `( output_buf_elems USIZE )`
+//!     * `( dict_buf_elems USIZE )`
+//! * Comment lines. These are any lines just containing a `( ... )` style forth comment.
+//! * Successful input lines, starting with `> ...`.
+//! * Successful output lines, starting with `< ...`.
+//!     * Any successful input line can have zero or more output lines
+//!     * If *no* input lines are specified, ANY successful output is accepted/ignored.
+//! * Unsuccessful input lines, starting with `x ...`.
+//!     * This line is expected to cause an "exception" - basically `process_line` returns
+//!       an `Err()`.
+//!     * There is no way to specify which error yet
+//!     * Unsuccessful input lines may not have any successful output
+//!
+//! These ui-tests can also be run as doctests (see below), and doctests can be run
+//! in miri.
+//!
+//! ### Example
+//!
+//! This is a forth ui-test doctest. It will be run with `cargo test --all-features`.
+//!
+//! ```rust
+//! # use forth3::testutil::blocking_runtest;
+//! #
+//! # blocking_runtest(r#"
+//! ( specify VM settings with frontmatter )
+//! ( data_stack_elems 1 )
+//!
+//! ( specify input with no output )
+//! > : star 42 emit ;
+//!
+//! ( specify input and output )
+//! > star
+//! < *ok.
+//!
+//! ( specify lines that cause exceptions/errors )
+//! x starb
+//! # "#)
+//! ```
+
+use crate::{leakbox::{LBForthParams, LBForth}, Forth, Error};
+
+/// Run the given forth ui test against ALL enabled forth VMs
+///
+/// A helper for calling blocking + async runtest functions. This is a good
+/// default to use for unit tests.
+pub fn all_runtest(contents: &str) {
+    blocking_runtest(contents);
+    #[cfg(feature = "async")]
+    async_blockon_runtest(contents);
+}
+
+/// Run the given forth ui test against the default forth vm
+///
+/// Does accept any/all/none of the following configuration frontmatter (see above
+/// for listing of frontmatter kinds)
+pub fn blocking_runtest(contents: &str) {
+    let tokd = tokenize(contents, true).unwrap();
+    let mut forth = LBForth::from_params(tokd.settings, (), Forth::FULL_BUILTINS);
+    blocking_steps_with(tokd.steps.as_slice(), &mut forth.forth);
+}
+
+/// Run the given forth ui-test against the given forth vm.
+///
+/// Does not accept ui-tests with frontmatter configuration (will panic)
+pub fn blocking_runtest_with<T>(contents: &str, forth: &mut Forth<T>) {
```

Comment:
nit, extremely bikesheddy: i might have put the test contents second, because it's the longer argument. IMO, this
```rust
blocking_runtest_with(forth, r#"
   ...
   ...
   ...
"#);
```
looks nicer than this:
```rust
blocking_runtest_with(r#"
       ...
       ...
       ...
    "#,
    forth,
);
```
similar to how closures are usually the last argument.

Reply:
oh that looks nice! I'll go with that.